### PR TITLE
Blog Post: Deferring Side-Effects in Node.js until the End of a Transaction

### PR DESCRIPTION
**Pull request by evantahler**
A post about our CLS work (from https://github.com/grouparoo/grouparoo/pull/1156) and automatic-transactionification.

TODO:
- [x] Do the examples make sense?  Are they too contrived?
- [x] Find a better image

### CODE REVIEW

File: pages/blog/defering-side-effects-in-node.mdx [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+<Image
+  centered
+  alt="Steps of a simple sync"
+  src="defering-side-effects-in-node/filiberto-santillan-HeyFNqApSLQ-unsplash.jpg"
+  width={1000}
+  height={667}
+/>
+
+At Grouparoo, we use [Actionhero](https://www.actionherojs.com) as our Node.js API server and [Sequelize](https://sequelize.org) for our Object Relational Mapping (ORM) tool - making it easy to work with complex records from our database. Within our Actions and Tasks, we often want to treat the whole execution as a singe [database transaction](https://stackoverflow.com/questions/974596/what-is-a-database-transaction) - either all the modifications to the database will succeed or fail as a unit. This is really helpful when a single activity may create or modify many database rows.
```
Comment (bleonard) [inline pages/blog/defering-side-effects-in-node.mdx]: single

File: pages/blog/defering-side-effects-in-node.mdx [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+## Why do we need Transactions?
+
+Take the following example from a prototypical blogging site. When a user is created (`POST /api/v1/user`), we also create their first post and send them a welcome Email. All examples in this post are written in Typescript, but the concepts work the same for Javascript.
```
Comment (bleonard) [inline pages/blog/defering-side-effects-in-node.mdx]: lowercase email

File: pages/blog/defering-side-effects-in-node.mdx [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+3. Sending the welcome email is not transaction safe.
+
+Sending the email as-written will happen even if we roll back the transaction because of an error when creating the new post… which isn’t great if the user record wasn’t committed! So what do we do?
+
+## Automatically Pass Transactions to all Queries: CLS-Hooked
+
+The solution to our problem comes from a wonderful package called [`cls-hooked`](https://github.com/Jeff-Lewis/cls-hooked). Using the magic of [`AsyncHooks`](https://github.com/nodejs/node/blob/master/doc/api/async_hooks.md), this package can tell when certain code is _within_ a callback chain or promise. In this way, you can say: "for all methods invoked within this async function, I want to keep this variable in scope". This is pretty wild! If you opt into using Sequelzie with CLS-Hooked, _every_ SQL statement will check to see if there is already a transaction in scope... You don't need to manually supply it as an argument!
```
Comment (bleonard) [inline pages/blog/defering-side-effects-in-node.mdx]: type: Sequelize

**Review by bleonard — COMMENTED** [review]
saw some typos

**Review by seancdavis — APPROVED** [review]

**Pull request merged**